corehq/apps/domain/forms.py

```diff
@@ -2377,3 +2373,88 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
                     css_class="disabled"
                 )
             )
+
+
+class ManageAppReleasesForm(forms.Form):
+    app_id = forms.ChoiceField(label=ugettext_lazy("Application"), choices=(), required=False)
+    location_id = forms.ChoiceField(label=ugettext_lazy("Location"), choices=(), required=False)
+    version = forms.IntegerField(label=ugettext_lazy('Version'), required=False, widget=Select(choices=[]))
+
+    def __init__(self, request, domain, *args, **kwargs):
+        self.domain = domain
+        super(ManageAppReleasesForm, self).__init__(*args, **kwargs)
+        self.fields['app_id'].choices = self.app_id_choices()
+        if request.GET.get('app_id'):
+            self.fields['app_id'].initial = request.GET.get('app_id')
+        self.fields['location_id'].choices = self.location_id_choices()
+        if request.GET.get('location_id'):
+            self.fields['location_id'].initial = request.GET.get('location_id')
+        if request.GET.get('version'):
+            self.fields['version'].initial = request.GET.get('version')
+        self.helper = HQFormHelper()
+        self.helper.form_tag = False
+
+        self.helper.layout = crispy.Layout(
+            crispy.Field('app_id', id='app-id-search-select', css_class="ko-select2"),
+            crispy.Field('location_id', id='location-search-select', css_class="ko-select2"),
+            crispy.Field('version', id='version-input'),
+            hqcrispy.FormActions(
+                crispy.ButtonHolder(
+                    crispy.Button('search', ugettext_lazy("Search"), data_bind="click: search"),
+                    crispy.Button('clear', ugettext_lazy("Clear"), data_bind="click: clear"),
+                    Submit('submit', ugettext_lazy("Add Release Restriction"))
+                )
+            )
+        )
+
+    def app_id_choices(self):
+        choices = [(None, _('Select Application'))]
+        for app in get_brief_apps_in_domain(self.domain):
+            choices.append((app.id, app.name))
+        return choices
+
+    def location_id_choices(self):
+        choices = [(None, _('Select Location'))]
+        choices.extend(SQLLocation.active_objects.filter(domain=self.domain).values_list('location_id', 'name'))
+        return choices
```

Comment:
`name` is not unique on a domain, so users won't be able to distinguish.  Elsewhere we use `location.display_name`, which includes the type, but my preference is `get_path_display`, which is guaranteed to be unique (each location name is unique among immediate siblings).

@orangejenny wrote
> The user is still going to see the list, it'd be good for that list to look sensical. Perhaps not worthwhile if it's expensive to sort, but if it's cheap, might as well do it.

The proper ordering for a path display is "tree order", either depth first or breadth first, not sure which is better.  I believe it happens automatically with a `get_descendants` query.  Probably out of scope of this PR, though.

More critical, however is the volume here.  IIRC, ICDS has (or will have) around 4 million locations, though you could lower that substantially by excluding the bottom couple levels.  That's probably too many to render in the html up-front.  Luckily, we do already have the server side of an ajax select2 widget for locations: https://github.com/dimagi/commcare-hq/blob/9bdfffdaf6f8a5a500cc5c4f9a01f99ce48f799b/corehq/apps/locations/urls.py#L25-L25
It's based on the "Groups or Users" filter options, but only pulls locations.  Might do some weird namespacing thing with the ids though, I don't recall.

Reply:
Thanks @esoergel 
Integrated in [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/412d18c72dcfe4cf5449bbd9d6f50853a5f054a3). Also updated the library to allow initial value in [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/f837c589a5a0316be798506052479b56b83e5f91)